samples/nrf9160/gps/src/main.c

```diff
@@ -130,85 +133,97 @@ static int activate_lte(bool activate)
 }
 #endif
 
-static int gnss_ctrl(uint32_t ctrl)
+static void gnss_event_handler(int event)
 {
 	int retval;
-
-	nrf_gnss_fix_retry_t    fix_retry    = 0;
-	nrf_gnss_fix_interval_t fix_interval = 1;
-	nrf_gnss_delete_mask_t	delete_mask  = 0;
-	nrf_gnss_nmea_mask_t	nmea_mask    = NRF_GNSS_NMEA_GSV_MASK |
-					       NRF_GNSS_NMEA_GSA_MASK |
-					       NRF_GNSS_NMEA_GLL_MASK |
-					       NRF_GNSS_NMEA_GGA_MASK |
-					       NRF_GNSS_NMEA_RMC_MASK;
-
-	if (ctrl == GNSS_INIT_AND_START) {
-		gnss_fd = nrf_socket(NRF_AF_LOCAL,
-				     NRF_SOCK_DGRAM,
-				     NRF_PROTO_GNSS);
-
-		if (gnss_fd >= 0) {
-			printk("GPS Socket created\n");
+	struct nrf_modem_gnss_nmea_data_frame *nmea_data;
+
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		nrf_modem_gnss_read(&last_pvt, sizeof(last_pvt), NRF_MODEM_GNSS_DATA_PVT);
+		new_data = true;
+		break;
+
+	case NRF_MODEM_GNSS_EVT_FIX:
+		got_fix = true;
+		fix_timestamp = k_uptime_get();
+		break;
+
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		nmea_data = k_malloc(sizeof(struct nrf_modem_gnss_nmea_data_frame));
+
+		retval = nrf_modem_gnss_read(nmea_data,
+					     sizeof(struct nrf_modem_gnss_nmea_data_frame),
+					     NRF_MODEM_GNSS_DATA_NMEA);
+		if (retval == 0) {
+			retval = k_msgq_put(&nmea_queue, &nmea_data, K_NO_WAIT);
 		} else {
-			printk("Could not init socket (err: %d)\n", gnss_fd);
-			return -1;
+			k_free(nmea_data);
 		}
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_FIX_RETRY,
-					&fix_retry,
-					sizeof(fix_retry));
-		if (retval != 0) {
-			printk("Failed to set fix retry value\n");
-			return -1;
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+#ifdef CONFIG_SUPL_CLIENT_LIB
+		retval = nrf_modem_gnss_read(&last_agps,
+					     sizeof(last_agps),
+					     NRF_MODEM_GNSS_DATA_AGPS_REQ);
+		if (retval == 0) {
+			agps_requested = true;
 		}
+#endif
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_FIX_INTERVAL,
-					&fix_interval,
-					sizeof(fix_interval));
-		if (retval != 0) {
-			printk("Failed to set fix interval value\n");
-			return -1;
-		}
+	case NRF_MODEM_GNSS_EVT_BLOCKED:
+		gnss_blocked = true;
+		break;
 
-		retval = nrf_setsockopt(gnss_fd,
-					NRF_SOL_GNSS,
-					NRF_SO_GNSS_NMEA_MASK,
-					&nmea_mask,
-					sizeof(nmea_mask));
-		if (retval != 0) {
-			printk("Failed to set nmea mask\n");
-			return -1;
-		}
+	case NRF_MODEM_GNSS_EVT_UNBLOCKED:
+		gnss_blocked = false;
+		break;
+
+	default:
+		break;
 	}
+}
+
+static int gnss_ctrl(uint32_t ctrl)
+{
+	int retval;
+
+	switch (ctrl) {
+	case GNSS_INIT_AND_START:
+		retval = nrf_modem_gnss_init();
+		retval |= nrf_modem_gnss_event_handler_set(gnss_event_handler);
```

Comment:
I think it would be beneficial to do error checks for each API call. I suspect this code will be copied around by users, so checking for each call will be easier to debug

Reply:
Agree.